docs/howtos/extracting_intermediates.rst

```diff
@@ -0,0 +1,204 @@
+-------------------------------
+
+This is a useful pattern for cases where it's clear in what particular
+way you want to split your submodules. Any submodule you expose in ``setup`` can be used directly. In the limit, you
+can define all submodules in ``setup`` and avoid using ``nn.compact`` altogether.
+
+.. testcode::
+
+  class CNN(nn.Module):
+    def setup(self):
+      self.features = Features()
+      self.classifier = Classifier()
+
+    def __call__(self, x):
+      x = self.features(x)
+      x = self.classifier(x)
+      return x
+
+  class Features(nn.Module):
+    @nn.compact
+    def __call__(self, x):
+      x = nn.Conv(features=32, kernel_size=(3, 3))(x)
+      x = nn.relu(x)
+      x = nn.avg_pool(x, window_shape=(2, 2), strides=(2, 2))
+      x = nn.Conv(features=64, kernel_size=(3, 3))(x)
+      x = nn.relu(x)
+      x = nn.avg_pool(x, window_shape=(2, 2), strides=(2, 2))
+      x = x.reshape((x.shape[0], -1))  # flatten
+      return x
+
+  class Classifier(nn.Module):
+    @nn.compact
+    def __call__(self, x):
+      x = nn.Dense(features=256)(x)
+      x = nn.relu(x)
+      x = nn.Dense(features=10)(x)
+      x = nn.log_softmax(x)
+      return x
+
+  @jax.jit
+  def init(key, x):
+    variables = CNN().init(key, x)
+    return variables['params']
+
+  @jax.jit
+  def features(params, x):
+    # Proposal #686 should allow for this alternative:
+    #   return CNN().features.apply({"params": params['features']})
+    return CNN().apply({"params": params}, x,
+      method=lambda module, x: module.features(x))
+
+  params = init(jax.random.PRNGKey(0), batch)
+
+  features(params, batch)
+
+
+Use ``nn.Sequential``
+---------------------
+
+You could also define ``CNN`` using a simple implementation of a ``Sequential`` combinator (this is quite common in more stateful approaches). This may be useful
+for very simple models and gives you arbitrary model
+surgery. But it can be very limiting -- if you even want to add one conditional, you are 
+forced to refactor away from ``nn.Sequential`` and structure
+your model more explicitly.
+
+.. testcode::
+
+  class Sequential(nn.Module):
+    submodules: Sequence[nn.Module]
+
+    def setup(self):
+      # Bind layers to `self` -- `__setattr__` gives submodules
+      # names and connects their variables through `self.variables`
+      self.layers = self.submodules
+
+    def __call__(self, x):
+      for layer in self.layers:
+        x = layer(x)
+      return x
+
+  def CNN():
+    return Sequential([
+      nn.Conv(features=32, kernel_size=(3, 3)),
+      nn.relu,
+      lambda x: nn.avg_pool(x, window_shape=(2, 2), strides=(2, 2)),
+      nn.Conv(features=64, kernel_size=(3, 3)),
+      nn.relu,
+      lambda x: nn.avg_pool(x, window_shape=(2, 2), strides=(2, 2)),
+      lambda x: x.reshape((x.shape[0], -1)),  # flatten
+      nn.Dense(features=256),
+      nn.relu,
+      nn.Dense(features=10),
+      nn.log_softmax,
+    ])
+
+  @jax.jit
+  def init(key, x):
```

Comment:
Nit: If you rename each module differently you only have to define `init` once if you provide the module name as an argument.